packages/grid/x-data-grid/src/components/columnHeaders/GridBaseColumnHeaders.tsx

```diff
@@ -23,8 +23,9 @@ const GridColumnHeadersRoot = styled('div', {
   slot: 'ColumnHeaders',
   overridesResolver: (props, styles) => styles.columnHeaders,
 })<{ ownerState: OwnerState }>({
-  position: 'relative',
-  overflow: 'hidden',
+  position: 'sticky',
+  top: 0,
+  zIndex: 2,
```

Comment:
Should be at least 3 to stay above the detail panel content:
https://deploy-preview-10059--material-ui-x.netlify.app/x/react-data-grid/master-detail/

https://github.com/mui/mui-x/assets/13808724/d6e6e2b5-7ee2-48c2-b9c8-763183283082